src/NHibernate/Engine/ActionQueue.cs

```diff
@@ -505,106 +507,172 @@ private class InsertActionSorter
+				{
+					if (ReferenceEquals(this, other))
+					{
+						return true;
+					}
+					if (other == null)
+					{
+						return false;
+					}
+					return EntityName == other.EntityName;
+				}
+
+				public override int GetHashCode()
+				{
+					return EntityName.GetHashCode();
+				}
+			}
+
+			// _orderedBatches and _entityNameBatchIdentifier replace _latestBatch from Hibernate implementation.
+			private readonly List<BatchIdentifier> _orderedBatches = new List<BatchIdentifier>();
+			private readonly Dictionary<string, BatchIdentifier> _entityNameBatchIdentifier = new Dictionary<string, BatchIdentifier>();
+			private readonly Dictionary<object, BatchIdentifier> _entityBatchIdentifier;
 
 			// the map of batch numbers to EntityInsertAction lists
-			private readonly Dictionary<int, List<EntityInsertAction>> _actionBatches = new Dictionary<int, List<EntityInsertAction>>();
+			private readonly Dictionary<BatchIdentifier, List<EntityInsertAction>> _actionBatches = new Dictionary<BatchIdentifier, List<EntityInsertAction>>();
 
 			public InsertActionSorter(ActionQueue actionQueue)
 			{
 				_actionQueue = actionQueue;
 
 				//optimize the hash size to eliminate a rehash.
-				_entityBatchNumber = new Dictionary<object, int>(actionQueue.insertions.Count + 1);
+				_entityBatchIdentifier = new Dictionary<object, BatchIdentifier>(actionQueue.insertions.Count + 1);
 			}
 
 			public void Sort()
 			{
 				// the list of entity names that indicate the batch number
 				foreach (EntityInsertAction action in _actionQueue.insertions)
 				{
-					// remove the current element from insertions. It will be added back later.
-					var entityName = action.EntityName;
-
 					// the entity associated with the current action.
 					var currentEntity = action.Instance;
 
-					var batchNumber = GetBatchNumber(action, entityName);
-					_entityBatchNumber[currentEntity] = batchNumber;
-					AddToBatch(batchNumber, action);
+					BatchIdentifier batchIdentifier;
+					if (!_entityNameBatchIdentifier.TryGetValue(action.EntityName, out batchIdentifier))
+					{
+						batchIdentifier = new BatchIdentifier(action.EntityName);
+						_entityNameBatchIdentifier[action.EntityName] = batchIdentifier;
+						_orderedBatches.Add(batchIdentifier);
+					}
+					AddParentChildEntityNames(action, batchIdentifier);
+					_entityBatchIdentifier.Add(currentEntity, batchIdentifier);
+					AddToBatch(batchIdentifier, action);
 				}
 				_actionQueue.insertions.Clear();
 
-				// now rebuild the insertions list. There is a batch for each entry in the name list.
-				for (var i = 0; i < _actionBatches.Count; i++)
+				for (var i = 0; i < _orderedBatches.Count; i++)
 				{
-					var batch = _actionBatches[i];
-					foreach (var action in batch)
+					var batchIdentifier = _orderedBatches[i];
+					var entityName = batchIdentifier.EntityName;
+
+					//Make sure that child entries are not before parents
+					for (var j = i - 1; j >= 0; j--)
```

Comment:
Still thinking about it, this sort is O(N²). N is the number of distinct entity classes involved in the flush. If this gets merged, we may document that session having a lot of distinct entity classes to flush at once would trigger a costly queries sort if ordering is enabled. 

In principle a comparison based sort is Ω(N log N) (cannot be better than that), which already does not scale well. So with our added constraint of not being able of comparing all elements, I do not think we can do better than this O(N²).

About comparing all elements, we may have two unrelated set of entity classes, 1: A < B < C and 2: D < E < F, requiring ordering among themselves but not caring about the other set. When comparing one entity from a set (B by example) to an entity class of the other set (let's take F), they should not be considered equals.  
Otherwise a comparison based sorting algorithm may encounter B == F == A while A must be < B, which would wreck the algorithm.

So for using a comparison based algorithm, we would have to detect each set of comparable entity classes in the batch, assign them a global ranking arbitrarily, and use it when comparing entity classes across sets: could be worth it on big sorts, but normally a session is not supposed to have a big list of entity classes to handle in the same flush. (Note that ado.batch_size does not limit the number of classes to order, as of my understanding. It only causes the resulting set of queries for a class to get split after ordering.)